src/marks/raster.js

```diff
@@ -0,0 +1,428 @@
+      if (C[b] === ex) C[b] = C[a] === ex ? C[c] : C[c] === ex ? C[a] : mix2(C[a], C[c]);
+      if (C[c] === ex) C[c] = C[b] === ex ? C[a] : C[a] === ex ? C[b] : mix2(C[b], C[a]);
+    }
+  }
+
+  // Interpolate the interior of all triangles with barycentric coordinates
+  const mix3 = F && mixer3(F, random);
+  const nepsilon = -1e-12; // tolerance for points that are on a triangle's edge
+  for (let i = 0; i < triangles.length; i += 3) {
+    const ta = triangles[i];
+    const tb = triangles[i + 1];
+    const tc = triangles[i + 2];
+    const Ax = points[2 * ta];
+    const Bx = points[2 * tb];
+    const Cx = points[2 * tc];
+    const Ay = points[2 * ta + 1];
+    const By = points[2 * tb + 1];
+    const Cy = points[2 * tc + 1];
+    const x1 = Math.min(Ax, Bx, Cx);
+    const x2 = Math.max(Ax, Bx, Cx);
+    const y1 = Math.min(Ay, By, Cy);
+    const y2 = Math.max(Ay, By, Cy);
+    const z = (By - Cy) * (Ax - Cx) + (Ay - Cy) * (Cx - Bx);
+    if (!z) continue;
+    const ia = index[ta];
+    const ib = index[tb];
+    const ic = index[tc];
+    for (let x = Math.floor(x1); x < x2; x++) {
+      for (let y = Math.floor(y1); y < y2; y++) {
+        if (x < 0 || x >= width || y < 0 || y >= height) continue;
+        const ga = ((By - Cy) * (x - Cx) + (y - Cy) * (Cx - Bx)) / z;
+        if (ga < nepsilon) continue;
+        const gb = ((Cy - Ay) * (x - Cx) + (y - Cy) * (Ax - Cx)) / z;
+        if (gb < nepsilon) continue;
+        const gc = 1 - ga - gb;
+        if (gc < nepsilon) continue;
+        const k = (x + width * y) << 2;
+        if (F) ({r, g, b} = rgb(color(mix3(F[ia], ga, F[ib], gb, F[ic], gc))));
+        if (FO) a = (ga * FO[ia] + gb * FO[ib] + gc * FO[ic]) * 255;
+        imageData[k + 0] = r;
+        imageData[k + 1] = g;
+        imageData[k + 2] = b;
+        imageData[k + 3] = a;
+      }
+    }
+  }
+
+  context2d.putImageData(image, 0, 0);
+}
+
+// TODO adaptive supersampling in areas of high variance?
+// TODO configurable iterations per sample (currently 1 + 2)
+// see https://observablehq.com/@observablehq/walk-on-spheres-precision
+function rasterizeWalk(canvas, index, {color}, {fill: F, fillOpacity: FO}, {x: X, y: Y}) {
+  const random = randomLcg(42); // TODO allow configurable rng?
+  const {width, height} = canvas;
+  const context2d = canvas.getContext("2d");
+  const image = context2d.createImageData(width, height);
+  const imageData = image.data;
+  let {r, g, b} = rgb(this.fill) ?? {r, g, b};
+  let a = (this.fillOpacity ?? 1) * 255;
+  const delaunay = Delaunay.from(
+    index,
+    (i) => X[i],
+    (i) => Y[i]
+  );
+  let k = 0;
+  for (let y = 0, i; y < height; y++) {
+    for (let x = 0; x < width; x++, k += 4) {
+      i = delaunay.find(x, y, i);
+      for (let j = 0, cx = x, cy = y; j < 2; ++j) {
+        const radius = Math.hypot(X[index[i]] - cx, Y[index[i]] - cy);
+        const angle = random() * 2 * Math.PI;
+        cx += Math.cos(angle) * radius;
+        cy += Math.sin(angle) * radius;
```

Comment:
```suggestion
        const dx = X[index[i]] - cx;
        const dy = Y[index[i]] - cy;
        const angle = random() * 2 * Math.PI;
        const ca = Math.cos(angle);
        const sa = Math.sin(angle);
        cx += ca * dx - sa * dy;
        cy += sa * dx + ca * dy;
```

would save 1 Math.hypot, but I'm not sure it's worth the effort—makes the code a bit more difficult to grasp.